breacharbiter.go

```diff
@@ -556,31 +718,69 @@ func (b *breachArbiter) breachObserver(contract *lnwallet.LightningChannel,
 			desc.SigHashes = hc
 			desc.InputIndex = inputIndex
 
-			return lnwallet.CommitSpendRevoke(b.wallet.Cfg.Signer, &desc, tx)
+			return lnwallet.CommitSpendRevoke(
+				b.wallet.Cfg.Signer, &desc, tx)
 		}
 
-		// Finally, with the two witness generation funcs created, we
-		// send the retribution information to the utxo nursery.
+		// Assemble the retribution information that parameterizes the
+		// construction of transactions required to correct the breach.
 		// TODO(roasbeef): populate htlc breaches
-		b.breachedContracts <- &retributionInfo{
+		retInfo := &retributionInfo{
 			commitHash: breachInfo.BreachTransaction.TxHash(),
 			chanPoint:  *chanPoint,
 
+			remoteIdentity: chanInfo.RemoteIdentity,
+			capacity:       chanInfo.Capacity,
+			settledBalance: chanInfo.LocalBalance,
+
 			selfOutput: &breachedOutput{
-				amt:         btcutil.Amount(localSignDesc.Output.Value),
-				outpoint:    breachInfo.LocalOutpoint,
-				witnessFunc: localWitness,
+				amt:            btcutil.Amount(localSignDesc.Output.Value),
+				outpoint:       breachInfo.LocalOutpoint,
+				signDescriptor: localSignDesc,
+				witnessType:    lnwallet.CommitmentNoDelay,
+				witnessFunc:    localWitness,
 			},
 
 			revokedOutput: &breachedOutput{
-				amt:         btcutil.Amount(remoteSignDesc.Output.Value),
-				outpoint:    breachInfo.RemoteOutpoint,
-				witnessFunc: remoteWitness,
+				amt:            btcutil.Amount(remoteSignDesc.Output.Value),
+				outpoint:       breachInfo.RemoteOutpoint,
+				signDescriptor: remoteSignDesc,
+				witnessType:    lnwallet.CommitmentRevoke,
+				witnessFunc:    remoteWitness,
 			},
 
+			htlcOutputs: []*breachedOutput{},
+
 			doneChan: make(chan struct{}),
 		}
 
+		// Persist the pending retribution state to disk.
+		if err := b.retributionStore.Add(retInfo); err != nil {
```

Comment:
Technically yes, a failure of either could cause the breach arbiter to enter a potentially invalid state. The issue here is that we currently don't have a mechanism for distinguishing fatal vs temporary errors. Perhaps the safest behavior in the meantime would be to only shutdown if we continually fail after some number of retries? Hopefully @Roasbeef can shed some wisdom here.